### PR TITLE
Improve tiling edge window previews: add "hover to activate" options

### LINKED ISSUE #779: Automatically scroll on hover
**Is your feature request related to a problem? Please describe.**
Not really, I don't have any problems using the Tiling WM

**Describe the solution you'd like**
When hovering the mouse at the end side of the screen, instead of having a preview and clicking that to go through the next window display, It would be nice for the window manager have an option to automatically scrolls to that next window upon hovering.

**Describe alternatives you've considered**
There's an existing gesture settings for touchpad users though. (swipe tiling windows options)

**Additional context**
image

### PR DESCRIPTION
**Pull request by jtaala**
Fixes #779.

This PR implements the following improvements to tiling edge window previews:
- fixes an issue where edge previews set stuck (don't disappear);
- fixes timing issues with continuous window (activation);
- adds an option to disable edge previews (simpler option compared to setting size to 0);
- adds an option to activate edge window by hover (see #779);
- adds an option to set the timeout in (ms) for hover activation;
- adds an option to enable/disable _continual activation of edge windows on hover_ (e.g. if pointer is still at edge).

image

### CODE REVIEW

**Comment by Thesola10** [comment]
Feels nice to use, but it doesn't seem to react while dragging a window, which is arguably an important use case

**Comment by Thesola10** [comment]
Also maybe require pushback or some other continued movement for consecutive scrolling so as to not cause unwanted movement just leaving the pointer at rest on screen edge

**Comment by jtaala** [comment]
> Also maybe require pushback or some other continued movement for consecutive scrolling so as to not cause unwanted movement just leaving the pointer at rest on screen edge

Oh lol, I specifically added that behaviour because it felt nice to leave the mouse there and continue to activate with a timeout...

I can put an option to enable/disable that behaviour I guess.

**Comment by jtaala** [comment]
> Feels nice to use, but it doesn't seem to react while dragging a window, which is arguably an important use case

This one will likely need to wait.  DnD's need a target to detect / activate clutter actors (basically same issue as https://github.com/paperwm/PaperWM/pull/574) - which means we'll need to use a PointerWatcher to detect the pointer being on the edge (when it's dragging a window) to activate an edge window.  

Doable, but means we need to rewrite `StackOverlay` approach to use a single `PointerWatcher` to detect pointer being at any monitor edge (instead of two clutter actors to detect pointer entry/exists).  That way, we can trigger preview (and timeout activations) - regardless of if we're dragging a window or not.

I think that's the issue anyway.  I'll get this PR in and have a look at the a `PointerWatcher` based implementation for this.

**Comment by jtaala** [comment]
> I think that's the issue anyway. I'll get this PR in and have a look at the a `PointerWatcher` based implementation for this.

@Thesola10, if you're interested, I created a quick implementation of pointerWatching edge positions.  It's responding to dragged windows (if you turn on activate on hover)... still not switching though so the dragging must be blocking the `Main.activateWindow()` call (or something else going on there): https://github.com/paperwm/PaperWM/commit/676defe5981f3ce5253a558a05ed5e36aaf82033

https://github.com/paperwm/PaperWM/tree/experimental-drag-drop-edge-window-previews

This breaks button press on preview (disabled) - so if just experimental / concept testing atm.

**Comment by jtaala** [comment]
> > I think that's the issue anyway. I'll get this PR in and have a look at the a `PointerWatcher` based implementation for this.
> 
> @Thesola10, if you're interested, I created a quick implementation of pointerWatching edge positions. It's responding to dragged windows (if you turn on activate on hover)... still not switching though so the dragging must be blocking the `Main.activateWindow()` call (or something else going on there): [676defe](https://github.com/paperwm/PaperWM/commit/676defe5981f3ce5253a558a05ed5e36aaf82033)
> 
> https://github.com/paperwm/PaperWM/tree/experimental-drag-drop-edge-window-previews
> 
> This breaks button press on preview (disabled) - so if just experimental / concept testing atm.

Anyways, some rethinking on concepts and implementations there - will have to tackle / look at that one at some point.  Won't be in this PR though.

**Comment by jtaala** [comment]
> Also maybe require pushback or some other continued movement for consecutive scrolling so as to not cause unwanted movement just leaving the pointer at rest on screen edge

Added an option to enable/disable this behaviour (if disabled, it won't activate consecutive activation - e.g. will have to move pointer away and back again... or click if you have _click to activate_ enabled).

**Comment by jtaala** [comment]
P.S. try that one out (with `continual activation`) - if still doesn't feel right (having to pull mouse back and retouch edge), then we can look at adding a `PressureBarrier` for continual activation.

**Comment by Thesola10** [comment]
Yeah I think I'd prefer a Pressure barrier, nonconsecutive just feels worse

Also experimental branch doesn't seem to track moving windows at all? idk

**Comment by jtaala** [comment]
> Also experimental branch doesn't seem to track moving windows at all? idk

Well, it's "working" in the sense that it overcomes the main issue with dragging windows and previews (detection of when the dragged window is at the edge).  Normally, you'll get no reaction or preview showing.  In experimental branch, you'll get this:

[Screencast from 2024-05-06 21-39-31.webm](https://github.com/paperwm/PaperWM/assets/30424662/ab9de42f-c3ac-4af4-8e29-36d45bd72325)

The flashing preview on the edge (when dragging) is just showing that it's detected and it's trying to activate the next window (but as I mentioned there's some other issues blocking that).  Still, it's just showing that using a PointerWatcher we can then detect when it's at an edge (whether dragging or not).

I.e. we'll have to rewrite `StackOverlay.js` to use a global pointerWatcher detects when it's at an edge - instead of current `StackOverlay` Clutter actors.

**Comment by jtaala** [comment]
Okay, all, have experimented with pressures barriers and mouse "pushback" to non-consecutive hover activation.  Pressure barriers don't work that well unless we move the mouse back a bit so the barrier resets.

In any case, I've now implemented mouse "pushback" for non-consecutive hover activations.  That is, after activation, we "push" the mouse back several pixels.  This means you can avoid accidental activations (by leaving the mouse on the edge), and if you want to show next preview, you can nudge the mouse against the edge again.

Consecutive (continual) is still an option in the settings.

**Comment by jtaala** [comment]
I do like the feel of this one (non-continual), even with a hover time of 0ms (instant activation).

**Comment by Thesola10** [comment]
Oh yeah, feels really nice with 0ms, lgtm

File: stackoverlay.js [hunk 1]
```diff
@@ -255,7 +347,24 @@ export class StackOverlay {
...
+        const scale = Settings.prefs.edge_preview_scale;
+        if (scale <= 0) {
+            return;
+        }
+
+        // if timeout is enabled, only show if valid timeout (e.g. if SHOW_DELAY is
+        /**
+         * if timeout is enabled, only show if valid timeout (e.g. if SHOW_DELAY is lte
+         * than timeout, then won't see the preview anyway).
+         */
```
Comment (Thesola10) [inline stackoverlay.js]:
```suggestion
        /**
         * if timeout is enabled, only show if valid timeout (e.g. if SHOW_DELAY <= timeout,
         * then won't see the preview anyway).
         */
```

quick comment cleanup

**Review by Thesola10 — APPROVED** [review]

**Comment by Thesola10** [comment]
think we could make the pushback a bit earlier so pressure action feels actually consecutive and doesn't wait for the end of the previous animation?

**Comment by jtaala** [comment]
> think we could make the pushback a bit earlier so pressure action feels actually consecutive and doesn't wait for the end of the previous animation?

Not sure I properly understand what you're suggesting (and how it would work), but feel free to play around with modifying this behaviour and seeing how it feels.  You'd probably want to change this: (which actions an edge-check and shows preview after animation time):

https://github.com/paperwm/PaperWM/blob/4deb4f730cfa0dc9ec2e7574790c0f5cadf31008/stackoverlay.js#L224-L235

**Pull request merged**